### ISSUE TITLE
Question about determinizing an FST

### ISSUE THREAD

**Issue by huangruizhe** [k2-fsa/k2#1040]
I have used k2 to generate the following FST:
![image](https://user-images.githubusercontent.com/28626513/184786915-3319e1cd-90f0-4a4d-9a9e-20dfeaa9e682.png)

It is supposed to be equivalent to this FST:
![image](https://user-images.githubusercontent.com/28626513/184788844-18c19775-0388-4254-8fed-4d5936895834.png)
And I think they are equivalent -- because in the first FST, if we merge state2 and state3, the result will become the second FST.

Then I tried to determinize the first FST (since state1 is not deterministic for input label "I"), but it will take a long time until memory runs out. Would you suggest what else I could do? Or perhaps the first FST does not meet some conditions for determinizability?
Thanks!

I am putting the codes here so people may reproduce the first FST.
```
s = ''' 
  0 1 5 0
  1 1 5 0
  1 2 1 0
  1 3 1 0
  2 2 1 0
  2 4 5 0
  3 3 1 0
  3 5 2 0
  4 4 5 0
  4 5 2 0
  5 5 2 0
  5 6 3 0
  6 6 3 0
  6 7 5 0
  7 7 5 0
  7 8 -1 0
  8
'''
fsa = k2.Fsa.from_str(s)
fsa.aux_labels = torch.tensor([ 5,  0,  1,  1,  0,  5,  0,  2,  0,  2,  0,  3,  0,  5,  0, -1], dtype=torch.int32)

labels_sym = k2.SymbolTable.from_str('''
            <blk> 0
            I 1
            _a 2
            m 3
            x 4
            <sil> 5
        ''')
aux_labels_sym = k2.SymbolTable.from_str('''
            I 1
            _a 2
            m 3
            x 4
            <sil> 5
        ''')

fsa.labels_sym = labels_sym
fsa.aux_labels_sym = aux_labels_sym

k2.to_dot(fsa)
```

Then, this will run into OOM:
```
fsa = k2.determinize(fsa)
```

**Comment by danpovey** [comment]
Hm, I'm afraid that in order to ensure the operation was differentiable, I implemented a different determinization algorithm that cannot differentiate that FST.  [In fact, there would be no clear way to differentiate that operation even if we had an algorithm that handled it.]
It may be better to use OpenFST for tasks like graph creation.
@csukuangfj what is the best entry point for that?

**Comment by csukuangfj** [comment]
> what is the best entry point for that?

@huangruizhe 
Maybe you can use https://github.com/csukuangfj/kaldifst to do that

I have not found time to add documentation to it yet, but it is usable. You can find example usages by looking at the tests.
See https://github.com/csukuangfj/kaldifst/tree/master/kaldifst/python/tests

As for the determinization, you can look at
- https://github.com/csukuangfj/kaldifst/blob/master/kaldifst/python/tests/test_determinize.py
- https://github.com/csukuangfj/kaldifst/blob/master/kaldifst/python/tests/test_determinize_star.py

**Comment by huangruizhe** [comment]
That's cool! Thanks for the detailed suggestions!

**Comment by csukuangfj** [comment]
@huangruizhe 
I have updated the doc for kaldifst. Please see https://k2-fsa.github.io/kaldifst/

![Screen Shot 2022-10-09 at 00 03 40](https://user-images.githubusercontent.com/5284924/194716659-d6f524d6-677d-49da-858a-735f68372e8c.png)

**Comment by huangruizhe** [comment]
Thanks for making the nice doc!! That's really clear and nice.
I've also learned to use kaldifst from the test cases. The doc will be more handy for sure.
